cpg_pipes/targets.py

```diff
@@ -0,0 +1,640 @@
+
+    @property
+    def target_id(self) -> str:
+        """
+        ID should be unique across target of all levels.
+
+        We are raising NotImplementedError instead of making it an abstract class, 
+        because mypy is not happy about binding TypeVar to abstract classes, see:
+        https://stackoverflow.com/questions/48349054/how-do-you-annotate-the-type-of
+        -an-abstract-class-with-mypy
+
+        Specifically,
+        ```
+        TypeVar('TargetT', bound=Target)
+        ```
+        Will raise:
+        ```
+        Only concrete class can be given where "Type[Target]" is expected
+        ```
+        """
+        raise NotImplementedError
+
+    def get_job_attrs(self) -> dict[str, str]:
+        """
+        Attributes for Hail Batch job.
+        """
+        raise NotImplementedError
+
+    def get_job_prefix(self) -> str:
+        """
+        Prefix job names.
+        """
+        raise NotImplementedError
+
+    def get_sequencing_type(self) -> SequencingType:
+        """
+        Sequencing type of samples. Will throw an error if more than
+        one sequencing type is found.
+        """
+        sequencing_types = set(s.sequencing_type for s in self.get_samples())
+        if len(sequencing_types) > 1:
+            raise ValueError(
+                f'Samples of more than one sequencing type used for joint calling: '
+                f'{sequencing_types}. Joint calling can be only performed on samples '
+                f'of the same type.'
+            )
+        return list(sequencing_types)[0]
+
+
+class Cohort(Target):
+    """
+    Represents a "cohort" target - all samples from all datasets in the pipeline.
+    """
+
+    def __init__(
+        self,
+        analysis_dataset_name: str,
+        namespace: Namespace,
+        name: str | None = None,
+        storage_provider: StorageProvider | None = None,
+    ):
+        """
+        @param name: name of the cohort
+        @param analysis_dataset_name: deferring creation of the analysis
+        dataset in case if it is also a part of the cohort datasets.
+        """
+        super().__init__()
+        self.name = name or analysis_dataset_name
+        self.namespace = namespace
+        self.storage_provider = storage_provider
+        self.analysis_dataset = Dataset(
+            name=analysis_dataset_name,
+            namespace=namespace,
+            storage_provider=self.storage_provider,
+        )
+        self._datasets_by_name: dict[str, Dataset] = {}
+
+    def __repr__(self):
+        return f'Cohort("{self.name}", {len(self.get_datasets())} datasets)'
+
+    @property
+    def target_id(self) -> str:
+        """Unique target ID"""
+        return self.name
+
+    def get_datasets(self, only_active: bool = True) -> list['Dataset']:
+        """
+        Gets list of all datasets.
+        Include only "active" datasets (unless only_active is False)
+        """
+        return [
```

Comment:
What's an inactive dataset? I might be missing context so feel free to ignore if it's obvious with other context.

Reply:
Base `Target` class defines `.active` field, which is inherited by all subclasses, so using it for consistency.
Also, in the pipelines logic, `active=False` is used in couple with `--skip-samples-with-missing-input` to flag targets (samples and datasets) that are not to be used in the pipeline:
https://github.com/populationgenomics/production-pipelines/blob/initial/cpg_pipes/pipeline/pipeline.py#L586-L587